File: sdks/oracles/rust/src/ix/proposals.rs
//! Proposal-round instruction builders (Ix 11–12).

use kassandra_oracles_program::instruction::Ix;
use solana_instruction::{AccountMeta, Instruction};
use solana_pubkey::Pubkey;

use super::build;
use crate::{SYSTEM_PROGRAM_ID, TOKEN_PROGRAM_ID};

// ===================================================================== Ix 11
/// `Propose` (Ix 11) — register a categorical `option` with a KASS `bond`.
#[allow(clippy::too_many_arguments)]
pub fn propose(
    program_id: &Pubkey,
    oracle: Pubkey,
    proposer: Pubkey,
    authority: Pubkey,
    authority_kass: Pubkey,
    stake_vault: Pubkey,
    option: u8,
    bond: u64,
) -> Instruction {
    let mut data = Vec::with_capacity(1 + 9);
    data.push(Ix::Propose as u8);
    data.push(option);
    data.extend_from_slice(&bond.to_le_bytes());
    build(
        program_id,
        vec![
            AccountMeta::new(oracle, false),
            AccountMeta::new(proposer, false),
            AccountMeta::new(authority, true),
            AccountMeta::new(authority_kass, false),
            AccountMeta::new(stake_vault, false),
            AccountMeta::new_readonly(TOKEN_PROGRAM_ID, false),
            AccountMeta::new_readonly(SYSTEM_PROGRAM_ID, false),
        ],
        data,
    )
}

// ===================================================================== Ix 12
/// `FinalizeProposals` (Ix 12) — close the proposal window. `tail` must be
/// EXACTLY `oracle.proposer_count` read-only Proposer accounts.
pub fn finalize_proposals(program_id: &Pubkey, oracle: Pubkey, tail: &[Pubkey]) -> Instruction {
    let mut accounts = Vec::with_capacity(1 + tail.len());
    accounts.push(AccountMeta::new(oracle, false));
    for p in tail {
        accounts.push(AccountMeta::new_readonly(*p, false));
    }
    build(program_id, accounts, vec![Ix::FinalizeProposals as u8])
}

File: sdks/oracles/ts/src/instructions/lifecycle/proposals.ts
/**
 * Proposal-flow lifecycle builders: `propose` (Ix=11), `finalizeProposals`
 * (Ix=12), `advancePhase` (Ix=7). See `../lifecycle` for conventions.
 */
import { Address, TransactionInstruction } from "@solana/web3.js";

import { Ix, KASSANDRA_PROGRAM_ID, SYSTEM_PROGRAM_ID, TOKEN_PROGRAM_ID } from "../../constants.js";
import * as pda from "../../pda.js";
import type { AddressInput } from "../../pda.js";
import { u64LE, u8, withDisc } from "../payload.js";
import { addr, ro, w } from "./shared.js";

// ---------------------------------------------------------------------------
// Propose (Ix=11) — processor/propose.rs
// Accounts: 0 oracle(w) 1 proposer(w,PDA) 2 authority(w,signer) 3 authority_base(w)
//           4 stake_vault(w,PDA) 5 token program(ro) 6 system program(ro).
// Payload (9): option u8 ++ bond u64.
// ---------------------------------------------------------------------------
export interface ProposeArgs {
  /** The oracle being proposed against. */
  oracle: AddressInput;
  /** Proposer authority (signer): funds rent + bond-transfer authority. */
  authority: AddressInput;
  /** Authority's SOL token account — the bond source. */
  authorityBase: AddressInput;
  /** Categorical option proposed (< oracle.options_count). */
  option: number;
  /** SOL bond escrowed into the stake vault (> 0). */
  bond: bigint | number;
  programId?: Address;
}

export async function propose(args: ProposeArgs): Promise<TransactionInstruction> {
  const programId = args.programId ?? KASSANDRA_PROGRAM_ID;
  const oracle = addr(args.oracle);
  const proposer = await pda.proposer(oracle, addr(args.authority), programId);
  const stakeVault = await pda.stakeVault(oracle, programId);

  const data = withDisc(Ix.Propose, u8(args.option), u64LE(args.bond));

  return new TransactionInstruction({
    programId,
    keys: [
      w(oracle),
      w(proposer.address),
      w(addr(args.authority), true),
      w(addr(args.authorityBase)),
      w(stakeVault.address),
      ro(TOKEN_PROGRAM_ID),
      ro(SYSTEM_PROGRAM_ID),
    ],
    data,
  });
}

// ---------------------------------------------------------------------------
// FinalizeProposals (Ix=12) — processor/finalize_proposals.rs
// Accounts: 0 oracle(w), then the FULL proposer set as a READ-ONLY tail.
// Payload: empty.
// ---------------------------------------------------------------------------
export interface FinalizeProposalsArgs {
  /** The oracle to finalize. */
  oracle: AddressInput;
  /** The FULL proposer-PDA set (exactly `proposer_count`), each read-only. */
  proposers: ReadonlyArray<AddressInput>;
  programId?: Address;
}

export async function finalizeProposals(
  args: FinalizeProposalsArgs,
): Promise<TransactionInstruction> {
  const programId = args.programId ?? KASSANDRA_PROGRAM_ID;
  return new TransactionInstruction({
    programId,
    keys: [w(addr(args.oracle)), ...args.proposers.map((p) => ro(addr(p)))],
    data: withDisc(Ix.FinalizeProposals),
  });
}

// ---------------------------------------------------------------------------
// AdvancePhase (Ix=7) — processor/advance_phase.rs
// Accounts: 0 oracle(w). Payload: empty. (Permissionless: no signer.)
// ---------------------------------------------------------------------------
export interface AdvancePhaseArgs {
  /** The oracle to tick FactProposal -> FactVoting. */
  oracle: AddressInput;
  programId?: Address;
}

export async function advancePhase(args: AdvancePhaseArgs): Promise<TransactionInstruction> {
  const programId = args.programId ?? KASSANDRA_PROGRAM_ID;
  return new TransactionInstruction({
    programId,
    keys: [w(addr(args.oracle))],
    data: withDisc(Ix.AdvancePhase),
  });
}
